### PR TITLE
Fix excessive stack use when storing HepMCProduct

### PR DESCRIPTION
**Pull request by Dr15Jones**
When storing the HepMCProduct, we need to avoid a fully connected GenParticle/GenVertex graph to keep ROOT's recursion in check. This can be done by not stroring GenVertex::m_particles_in. We then refill that container in a ROOT ioread rule.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
This depends on https://github.com/cms-sw/cmsdist/pull/4745

**Comment by Dr15Jones** [comment]
@fabiocos this fix and the corresponding fix in ROOT is actually needed in all releases where we write HepMCProduct.

**Comment by Dr15Jones** [comment]
please test

**Comment by fabiocos** [comment]
@Dr15Jones as we write HepMCProduct virtually in any release (that is used for GEN production), I think we should decide where this is really relevant besides master: 10_5_X for sure, 10_2_X (in the middle of the production, but we have seen the problems appearing there). May be 10_3_X. What about older cycles?

File: SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include <HepMC/GenParticle.h>
```
Comment (fabiocos) [inline SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc]: @Dr15Jones although there is nothing wrong with it, I wonder whether calling this source as the namespace might not be a bit misleading

**Comment by fabiocos** [comment]
@efeyazgan @qliphy @alberto-sanchez this is a proposed fix for issues observed in production, it would be useful to get it in 10_6_0_pre1, unless you have comments about it

**Comment by Dr15Jones** [comment]
The problem was encountered in CMSSW_9_4_7.

File: SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc (hunk 1 above)
Comment (Dr15Jones) [reply SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc]: Naming suggestions are welcome.

Comment (fabiocos) [reply SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc]: may be HepMC_IOhelper ? I have no strong feeling about it

Comment (Dr15Jones) [reply SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc]: I think I prefer what is already there. That file does contain the only functions in that namespace.

**Comment by fabiocos** [comment]
@Dr15Jones in 9_4_X there is the latest one that you investigated as mentioned in #26022, right, I was trying to find back the original thread. So it would be good to go back at least up to 9_4_X. BTW, should we expect that this fix has also an impact on the recent problems seen in the test of 10_5_0_pre1_ROOT614, as discussed in #25919?

File: SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc (hunk 1 above)
Comment (fabiocos) [reply SimDataFormats/GeneratorProducts/src/hepmc_rootio.cc]: fine for me

**Comment by Dr15Jones** [comment]
>  BTW, should we expect that this fix has also an impact on the recent problems seen in the test of 10_5_0_pre1_ROOT614, as discussed in #25919?

yes

**Comment by fabiocos** [comment]
that is another build that needs to be done with this fix....

**Comment by Dr15Jones** [comment]
hold
I just did a more complete test and things did not work.

**Comment by Dr15Jones** [comment]
untold

**Comment by Dr15Jones** [comment]
unhold

**Comment by Dr15Jones** [comment]
My recent test of this pull request in CMSSW_10_6_X_2019-03-07-1200 shows that the recent fix to ROOT now allows this change to function. 

I can confirm that without this change, reading the original file causes a stack overflow and with this change I can read the file just fine. Surprising (to me at least) is this change allows reading both the newly created files AND the old ones!

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
@alberto-sanchez @efeyazgan @perrozzi @qliphy please review. This technical change is needed for both the switch to ROOT 6.14 but also to ongoing production (where it would have to be backported to 9_4).

**Comment by alberto-sanchez** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
@Dr15Jones it looks like there is plenty of errors in the ROOT614_X build after the merge of this PR:
https://cmssdt.cern.ch/SDT/html/cmssdt-ib/#/relVal/CMSSW_10_6/2019-03-10-0000?selectedArchs=slc7_amd64_gcc700&selectedFlavors=ROOT614_X&selectedStatus=failed

----- Begin Fatal Exception 10-Mar-2019 18:03:26 CET-----------------------
An exception of category 'WrongReference' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 2 stream: 3
   [1] Running path 'digitisation_step'
   [2] Calling method for module GenParticleProducer/'genParticles'
Exception Message:
barcodes are duplicated! 
----- End Fatal Exception -------------------------------------------------
GenEvent::delete_all_vertices strange result ... after deleting all vertices, 
the particle and vertex maps aren't empty.
  This probably indicates deeper problems or memory leak in the code.
Number vtx,particle the event after deleting = 0  136

**Comment by Dr15Jones** [comment]
@fabiocos @smuzaffar I'm pretty sure the ROOT 6.14 builds do not have the needed ROOT fix yet.

**Pull request merged**